web/book/src/internals/modules.md

```diff
@@ -0,0 +1,95 @@
+# Modules
+
+PRQL needs modules that would:
+
+- allow importing declarations from other files,
+- have namespaces for things like `std`,
+- have hierarchical structure so we can represent files in directories,
+- have unambiguous module structure within a project,
+- be able to compile individual files that are a part of a project.
+
+## Basic definition
+
+Module is a namespace that contains declarations.
+
+Modules can be defined with mod keyword and a code block encased in curly
+braces:
+
+```
+mod my_module {
+    let bangers = (from tracks | take 10)
+}
+```
+
+## File loading
```

Comment:
This design is heavily plagiarized from how Rust modules work. Thank god for opensource!

Reply:
I think Rust is poor choice to draw inspiration from for a module system. In my opinion it is very confusing and one of the worst parts of Rust.
I think the module system should aim to be as simple and predictable as possible.
I think it would be better to look at PHP and it's [`include`](https://www.php.net/manual/en/function.include.php) statement which is dead simple.

Reply:
I tried to redesign it to come closer to PHP's `include()` or Common.js's `require()` and I've realized that this is a very close already.

The main distinctions are:
- we have `module hello` declaration instead of a *function* `include("hello.prql")`,
- we have more restrictions on importing that make the module structure more "normalized" (see goal 4),
- these restrictions also allow cyclic references between modules,
- child modules don't have to re-import modules that are already imported into parent modules:
  ```
  # parent.prql
  module child1
  module child2
  ```
  ```
  # child1.prql

  # no need to re-import child2, as it is already in the module structure
  let _ = (from child2.some_decl | ...)
  ```